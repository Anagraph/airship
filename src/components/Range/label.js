import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { colors } from '../../constants';

const StyledLabel = styled.span`
  color: ${colors.type01};
  font: 500 10px/12px 'Roboto';
  transform: translateZ(0) translateX(-50%);
  white-space: nowrap;
  position: absolute;
  top: 10px;

  .is-disabled & {
    color: ${colors.ui03};
  }
`;

const Label = props => {
  const labelValue = props.formatLabel
    ? props.formatLabel(props.children, props.type)
    : props.children;

  return <StyledLabel>{labelValue}</StyledLabel>;
};

Label.propTypes = {
  children: PropTypes.node.isRequired,
  formatLabel: PropTypes.func,
  type: PropTypes.string.isRequired,
};

export default Label;
