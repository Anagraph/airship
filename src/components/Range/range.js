import React, { Component } from 'react';
import PropTypes from 'prop-types';
import * as valueTransformer from './transformer';
import Slider from './slider';
import Track from './track';
import {
  capitalize,
  distanceTo,
  isDefined,
  isObject,
  length
} from '../../utils';
import styled from 'styled-components';

const StyledRange = styled.div`
  position: relative;
  width: ${(props) => `${props.width}px`};
`;

/**
 * A React component that allows users to input numeric values within a range
 * by dragging its sliders.
 */
class Range extends Component {
  state = {
    value: this.isMultiValue()
      ? this.props.value
      : {
          min: this.props.minValue,
          max: this.props.value || this.props.minValue
        }
  };

  constructor(props) {
    super(props);

    this.startValue = null;
    this.node = null;
    this.trackNode = null;
    this.isSliderDragging = false;
    this.lastKeyMoved = null;
  }

  componentWillUnmount() {
    this.removeDocumentMouseUpListener();
    this.removeDocumentTouchEndListener();
  }

  getTrackClientRect() {
    return this.trackNode.getClientRect();
  }

  getKeyByPosition(position) {
    const positions = valueTransformer.getPositionsFromValues(
      this.state.value,
      this.props.minValue,
      this.props.maxValue,
      this.getTrackClientRect()
    );

    if (this.isMultiValue()) {
      const distanceToMin = distanceTo(position, positions.min);
      const distanceToMax = distanceTo(position, positions.max);

      if (distanceToMin < distanceToMax) {
        return 'min';
      }
    }

    return 'max';
  }

  getKeys() {
    if (this.isMultiValue()) {
      return ['min', 'max'];
    }

    return ['max'];
  }

  hasStepDifference(values) {
    const currentValues = this.state.value;
    return (
      length(values.min, currentValues.min) >= this.props.step ||
      length(values.max, currentValues.max) >= this.props.step
    );
  }

  isMultiValue() {
    return isObject(this.props.value);
  }

  isWithinRange(values) {
    if (this.isMultiValue()) {
      return (
        values.min >= this.props.minValue &&
        values.max <= this.props.maxValue &&
        values.min < values.max
      );
    }

    return (
      values.max >= this.props.minValue && values.max <= this.props.maxValue
    );
  }

  shouldUpdate(values) {
    return this.isWithinRange(values) && this.hasStepDifference(values);
  }

  updatePosition(key, position) {
    const positions = valueTransformer.getPositionsFromValues(
      this.state.value,
      this.props.minValue,
      this.props.maxValue,
      this.getTrackClientRect()
    );

    positions[key] = position;
    this.lastKeyMoved = key;

    this.updatePositions(positions);
  }

  updatePositions(positions) {
    const values = {
      min: valueTransformer.getValueFromPosition(
        positions.min,
        this.props.minValue,
        this.props.maxValue,
        this.getTrackClientRect()
      ),
      max: valueTransformer.getValueFromPosition(
        positions.max,
        this.props.minValue,
        this.props.maxValue,
        this.getTrackClientRect()
      )
    };

    const transformedValues = {
      min: valueTransformer.getStepValueFromValue(values.min, this.props.step),
      max: valueTransformer.getStepValueFromValue(values.max, this.props.step)
    };

    this.updateValues(transformedValues);
  }

  updateValue(key, value) {
    const values = this.state.value;
    values[key] = value;

    this.updateValues(values);
  }

  updateValues(values) {
    if (!this.shouldUpdate(values)) {
      return;
    }
    const { onChange } = this.props;

    this.setState(
      (state) => {
        return { value: values };
      },
      () => {
        onChange && onChange(this.isMultiValue() ? values : values.max);
      }
    );
  }

  addDocumentMouseUpListener() {
    this.removeDocumentMouseUpListener();
    this.node.ownerDocument.addEventListener('mouseup', this.handleMouseUp);
  }

  addDocumentTouchEndListener() {
    this.removeDocumentTouchEndListener();
    this.node.ownerDocument.addEventListener('touchend', this.handleTouchEnd);
  }

  removeDocumentMouseUpListener() {
    this.node.ownerDocument.removeEventListener('mouseup', this.handleMouseUp);
  }

  removeDocumentTouchEndListener() {
    this.node.ownerDocument.removeEventListener(
      'touchend',
      this.handleTouchEnd
    );
  }

  handleSliderDrag = (event, key) => {
    if (this.props.disabled) {
      return;
    }

    const position = valueTransformer.getPositionFromEvent(
      event,
      this.getTrackClientRect()
    );
    this.isSliderDragging = true;
    requestAnimationFrame(() => this.updatePosition(key, position));
  };

  handleTrackDrag = (event, prevEvent) => {
    if (this.props.disabled || !this.props.draggable || this.isSliderDragging) {
      return;
    }

    const { maxValue, minValue } = this.props;
    const { value: { max, min } } = this.state;

    const position = valueTransformer.getPositionFromEvent(
      event,
      this.getTrackClientRect()
    );
    const value = valueTransformer.getValueFromPosition(
      position,
      minValue,
      maxValue,
      this.getTrackClientRect()
    );
    const stepValue = valueTransformer.getStepValueFromValue(
      value,
      this.props.step
    );

    const prevPosition = valueTransformer.getPositionFromEvent(
      prevEvent,
      this.getTrackClientRect()
    );
    const prevValue = valueTransformer.getValueFromPosition(
      prevPosition,
      minValue,
      maxValue,
      this.getTrackClientRect()
    );
    const prevStepValue = valueTransformer.getStepValueFromValue(
      prevValue,
      this.props.step
    );

    const offset = prevStepValue - stepValue;

    const transformedValues = {
      min: min - offset,
      max: max - offset
    };

    this.updateValues(transformedValues);
  };

  handleTrackMouseDown = (event, position) => {
    if (this.props.disabled) {
      return;
    }

    const { maxValue, minValue } = this.props;
    const { value: { max, min } } = this.state;

    event.preventDefault();

    const value = valueTransformer.getValueFromPosition(
      position,
      minValue,
      maxValue,
      this.getTrackClientRect()
    );
    const stepValue = valueTransformer.getStepValueFromValue(
      value,
      this.props.step
    );

    if (!this.props.draggable || stepValue > max || stepValue < min) {
      this.updatePosition(this.getKeyByPosition(position), position);
    }
  };

  handleInteractionStart = () => {
    const { onChangeStart, onChangeComplete } = this.props;
    onChangeStart && onChangeStart(this.state.value);

    if (onChangeComplete && !isDefined(this.startValue)) {
      this.startValue = this.state.value;
    }
  };

  handleInteractionEnd = () => {
    const { onChangeComplete } = this.props;
    if (this.isSliderDragging) {
      this.isSliderDragging = false;
    }

    if (!onChangeComplete || !isDefined(this.startValue)) {
      return;
    }

    if (this.startValue !== this.state.value) {
      onChangeComplete(this.state.value);
    }

    this.startValue = null;
  };

  handleMouseDown = (event) => {
    this.handleInteractionStart(event);
    this.addDocumentMouseUpListener();
  };

  handleMouseUp = (event) => {
    this.handleInteractionEnd(event);
    this.removeDocumentMouseUpListener();
  };

  handleTouchStart = (event) => {
    this.handleInteractionStart(event);
    this.addDocumentTouchEndListener();
  };

  handleTouchEnd = (event) => {
    this.handleInteractionEnd(event);
    this.removeDocumentTouchEndListener();
  };

  renderSliders() {
    const values = this.state.value;
    const percentages = valueTransformer.getPercentagesFromValues(
      this.state.value,
      this.props.minValue,
      this.props.maxValue
    );
    const keys = this.getKeys();

    return keys.map((key) => {
      const value = values[key];
      const percentage = percentages[key];

      let { maxValue, minValue } = this.props;

      if (key === 'min') {
        maxValue = values.max;
      } else {
        minValue = values.min;
      }

      const slider = (
        <Slider
          classNames={this.props.classNames}
          formatLabel={this.props.formatLabel}
          key={key}
          maxValue={maxValue}
          minValue={minValue}
          onSliderDrag={this.handleSliderDrag}
          percentage={percentage}
          type={key}
          value={value}
        />
      );

      return slider;
    });
  }

  renderHiddenInputs() {
    if (!this.props.name) {
      return [];
    }

    const isMultiValue = this.isMultiValue();
    const values = this.state.value;

    return this.getKeys().map((key) => {
      const value = values[key];
      const name = isMultiValue
        ? `${this.props.name}${capitalize(key)}`
        : this.props.name;

      return <input key={key} type="hidden" name={name} value={value} />;
    });
  }

  render() {
    const { width, disabled, draggable } = this.props;
    const isMultiValue = this.isMultiValue();
    const percentages = valueTransformer.getPercentagesFromValues(
      this.state.value,
      this.props.minValue,
      this.props.maxValue
    );

    return (
      <StyledRange
        className={disabled ? 'is-disabled' : null}
        width={width}
        innerRef={(node) => {
          this.node = node;
        }}
        onMouseDown={this.handleMouseDown}
        onTouchStart={this.handleTouchStart}
      >
        <Track
          draggable={isMultiValue ? draggable : false}
          ref={(trackNode) => {
            this.trackNode = trackNode;
          }}
          percentages={percentages}
          onTrackDrag={this.handleTrackDrag}
          onTrackMouseDown={this.handleTrackMouseDown}
        >
          {this.renderSliders()}
        </Track>
        {this.renderHiddenInputs()}
      </StyledRange>
    );
  }
}

Range.defaultProps = {
  disabled: false,
  maxValue: 10,
  minValue: 0,
  value: 0,
  step: 1,
  width: 200
};

Range.propTypes = {
  disabled: PropTypes.bool,
  draggable: PropTypes.bool,
  formatLabel: PropTypes.func,
  maxValue: PropTypes.number,
  minValue: PropTypes.number,
  name: PropTypes.string,
  onChangeStart: PropTypes.func,
  onChange: PropTypes.func,
  onChangeComplete: PropTypes.func,
  step: PropTypes.number,
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.object]),
  width: PropTypes.number
};

export default Range;
