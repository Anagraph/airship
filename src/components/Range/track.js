import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { theme } from '../../constants';

const TrackBack = styled.div`
  background: ${props => props.theme.ui03};
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  user-select: none;

  .is-disabled & {
    background: ${props => props.theme.ui03};
    cursor: not-allowed;
  }
`;
TrackBack.defaultProps = {
  theme,
};

const TrackFront = TrackBack.extend`
  background: ${props => props.theme.brand01};
  transition: left 0.1s ease, width 0.1s ease;
  -webkit-tap-highlight-color: transparent;
  user-select: none;

  .is-disabled & {
    background: ${props => props.theme.ui03};
  }
`;
TrackFront.defaultProps = {
  theme,
};

class Track extends Component {
  constructor(props) {
    super(props);

    this.node = null;
    this.trackDragEvent = null;
  }

  getClientRect = () => this.node.getBoundingClientRect();

  getActiveTrackStyle = () => {
    const { percentages } = this.props;
    const { min, max } = percentages;
    const width = `${(max - min) * 100}%`;
    const left = `${min * 100}%`;
    return { left, width };
  };

  addDocumentMouseMoveListener() {
    this.removeDocumentMouseMoveListener();
    this.node.ownerDocument.addEventListener('mousemove', this.handleMouseMove, { pasive: false });
  }

  addDocumentMouseUpListener() {
    this.removeDocumentMouseUpListener();
    this.node.ownerDocument.addEventListener('mouseup', this.handleMouseUp);
  }

  removeDocumentMouseMoveListener() {
    this.node.ownerDocument.removeEventListener(
      'mousemove',
      this.handleMouseMove
    );
  }

  removeDocumentMouseUpListener() {
    this.node.ownerDocument.removeEventListener('mouseup', this.handleMouseUp);
  }

  handleMouseMove = event => {
    event.preventDefault();

    const { draggable, onTrackDrag } = this.props;
    if (!draggable) {
      return;
    }

    if (this.trackDragEvent !== null) {
      onTrackDrag(event, this.trackDragEvent);
    }

    this.trackDragEvent = event;
  };

  handleMouseUp = () => {
    if (!this.props.draggable) {
      return;
    }

    this.removeDocumentMouseMoveListener();
    this.removeDocumentMouseUpListener();
    this.trackDragEvent = null;
  };

  handleMouseDown = event => {
    const { draggable, onTrackMouseDown } = this.props;
    const clientX = event.touches ? event.touches[0].clientX : event.clientX;
    const trackClientRect = this.getClientRect();
    const position = {
      x: clientX - trackClientRect.left,
      y: 0,
    };

    onTrackMouseDown(event, position);

    if (draggable) {
      this.addDocumentMouseMoveListener();
      this.addDocumentMouseUpListener();
    }
  };

  handleTouchStart = event => {
    event.preventDefault();

    this.handleMouseDown(event);
  };

  render() {
    const activeTrackStyle = this.getActiveTrackStyle();

    return (
      <TrackBack
        onMouseDown={this.handleMouseDown}
        onTouchStart={this.handleTouchStart}
        innerRef={node => {
          this.node = node;
        }}
      >
        <TrackFront style={activeTrackStyle} />
        {this.props.children}
      </TrackBack>
    );
  }
}

Track.propTypes = {
  children: PropTypes.node,
  draggable: PropTypes.bool,
  onTrackDrag: PropTypes.func,
  onTrackMouseDown: PropTypes.func.isRequired,
  percentages: PropTypes.objectOf(PropTypes.number).isRequired,
};

export default Track;
