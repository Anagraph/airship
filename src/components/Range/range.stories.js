import React from 'react';
import { storiesOf } from '@storybook/react';
import { ThemeProvider } from 'styled-components';
import { action } from '@storybook/addon-actions';
import Range from './range';
import { theme } from '../../constants';

const CUSTOM_THEME = {
  ...theme,
  brand01: '#FABADA',
};

storiesOf('Range', module)
  .add('Default', () => (
    <div style={{ padding: '1rem' }}>
      <h3 className="header">Default</h3>
      <Range />

      <h3 className="header">Default disabled</h3>
      <Range disabled />

      <h3 className="header">With an event</h3>
      <Range onChange={action('changed')} />

      <h3 className="header">With multivalue</h3>
      <Range
        draggable
        value={{ min: 5, max: 15 }}
        minValue={0}
        maxValue={30}
      />
      <h3 className="header">With multivalue disabled</h3>
      <Range
        disabled
        draggable
        value={{ min: 5, max: 15 }}
        minValue={0}
        maxValue={30}
      />
    </div>
  ))
  .add('With custom theme', () => (
    <ThemeProvider theme={CUSTOM_THEME}>
      <Range />
    </ThemeProvider>
  ));
