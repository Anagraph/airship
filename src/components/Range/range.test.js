import React from 'react';
import Range from './range';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';

describe('Range', () => {
  it('renders without crashing', () => {
    const component = renderer.create(<Range />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('maxValue', () => {
    const component = renderer.create(<Range maxValue={20} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('minValue', () => {
    const component = renderer.create(<Range minValue={10} maxValue={30} />);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('value', () => {
    const component = renderer.create(
      <Range value={20} minValue={10} maxValue={30} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('disabled', () => {
    const component = renderer.create(
      <Range disabled value={20} minValue={10} maxValue={30} />
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('multiple sliders', () => {
    const component = mount(
      <Range disabled value={{ min: 5, max: 15 }} minValue={10} maxValue={30} />
    );

    expect(component.find('Slider').length).toBe(2);
  });

  it('onChange callback', () => {
    const spy = jest.fn();
    const component = renderer.create(
      <Range onChange={spy} value={20} minValue={10} maxValue={30} />
    );
    component.getInstance().updateValues({ min: 10, max: 15 });
    expect(spy).toHaveBeenCalled();
  });
});
