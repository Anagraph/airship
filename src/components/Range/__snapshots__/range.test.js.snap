// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Range disabled 1`] = `
.c5 {
  color: #E2E6E3;
  font: 500 10px/12px 'Roboto';
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  margin-top: 4px;
}

.c3 {
  left: 50%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: not-allowed;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  -webkit-transition: left 0.1s ease;
  transition: left 0.1s ease;
  top: -6px;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  padding: 0 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  -webkit-tap-highlight-color: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff;
  border-radius: 100%;
  border: 1px solid #1785FB;
  display: inline-block;
  height: 12px;
  outline: none;
  -webkit-transition: -webkit-transform 0.1s ease;
  -webkit-transition: transform 0.1s ease;
  transition: transform 0.1s ease;
  width: 12px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #F5F5F5;
  border: 1px solid #E2E6E3;
}

.c4:hover {
  -webkit-transform: scale(1.3);
  -ms-transform: scale(1.3);
  transform: scale(1.3);
}

.c4:hover {
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.c1 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #E2E6E3;
  cursor: not-allowed;
}

.c2 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #E2E6E3;
  cursor: not-allowed;
  background: #1785FB;
  -webkit-transition: left 0.1s ease,width 0.1s ease;
  transition: left 0.1s ease,width 0.1s ease;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  left: 0%;
  width: 50%;
  background: #D1D5D7;
}

.c0 {
  position: relative;
  padding: 8px 0 24px;
}

<div
  className="c0"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c1"
    disabled={true}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="c2"
      disabled={true}
      width="50%"
    />
    <div
      className="c3"
      disabled={true}
      draggable={undefined}
      onMouseDown={[Function]}
      onTouchMove={[Function]}
    >
      <span
        className="c4"
        disabled={true}
        tabIndex="0"
      />
      <span
        className="c5"
        disabled={true}
      >
        20
      </span>
    </div>
  </div>
</div>
`;

exports[`Range maxValue 1`] = `
.c5 {
  color: #2C2C2C;
  font: 500 10px/12px 'Roboto';
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  margin-top: 4px;
}

.c3 {
  left: 0%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  -webkit-transition: left 0.1s ease;
  transition: left 0.1s ease;
  top: -6px;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  padding: 0 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  -webkit-tap-highlight-color: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff;
  border-radius: 100%;
  border: 1px solid #1785FB;
  display: inline-block;
  height: 12px;
  outline: none;
  -webkit-transition: -webkit-transform 0.1s ease;
  -webkit-transition: transform 0.1s ease;
  transition: transform 0.1s ease;
  width: 12px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4:hover {
  -webkit-transform: scale(1.3);
  -ms-transform: scale(1.3);
  transform: scale(1.3);
}

.c1 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #1785FB;
  -webkit-transition: left 0.1s ease,width 0.1s ease;
  transition: left 0.1s ease,width 0.1s ease;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  left: 0%;
  width: 0%;
}

.c0 {
  position: relative;
  padding: 8px 0 24px;
}

<div
  className="c0"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c1"
    disabled={false}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="c2"
      disabled={false}
      width="0%"
    />
    <div
      className="c3"
      disabled={false}
      draggable={undefined}
      onMouseDown={[Function]}
      onTouchMove={[Function]}
    >
      <span
        className="c4"
        disabled={false}
        tabIndex="0"
      />
      <span
        className="c5"
        disabled={false}
      >
        0
      </span>
    </div>
  </div>
</div>
`;

exports[`Range minValue 1`] = `
.c5 {
  color: #2C2C2C;
  font: 500 10px/12px 'Roboto';
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  margin-top: 4px;
}

.c3 {
  left: 0%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  -webkit-transition: left 0.1s ease;
  transition: left 0.1s ease;
  top: -6px;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  padding: 0 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  -webkit-tap-highlight-color: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff;
  border-radius: 100%;
  border: 1px solid #1785FB;
  display: inline-block;
  height: 12px;
  outline: none;
  -webkit-transition: -webkit-transform 0.1s ease;
  -webkit-transition: transform 0.1s ease;
  transition: transform 0.1s ease;
  width: 12px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4:hover {
  -webkit-transform: scale(1.3);
  -ms-transform: scale(1.3);
  transform: scale(1.3);
}

.c1 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #1785FB;
  -webkit-transition: left 0.1s ease,width 0.1s ease;
  transition: left 0.1s ease,width 0.1s ease;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  left: 0%;
  width: 0%;
}

.c0 {
  position: relative;
  padding: 8px 0 24px;
}

<div
  className="c0"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c1"
    disabled={false}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="c2"
      disabled={false}
      width="0%"
    />
    <div
      className="c3"
      disabled={false}
      draggable={undefined}
      onMouseDown={[Function]}
      onTouchMove={[Function]}
    >
      <span
        className="c4"
        disabled={false}
        tabIndex="0"
      />
      <span
        className="c5"
        disabled={false}
      >
        10
      </span>
    </div>
  </div>
</div>
`;

exports[`Range renders without crashing 1`] = `
.c5 {
  color: #2C2C2C;
  font: 500 10px/12px 'Roboto';
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  margin-top: 4px;
}

.c3 {
  left: 0%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  -webkit-transition: left 0.1s ease;
  transition: left 0.1s ease;
  top: -6px;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  padding: 0 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  -webkit-tap-highlight-color: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff;
  border-radius: 100%;
  border: 1px solid #1785FB;
  display: inline-block;
  height: 12px;
  outline: none;
  -webkit-transition: -webkit-transform 0.1s ease;
  -webkit-transition: transform 0.1s ease;
  transition: transform 0.1s ease;
  width: 12px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4:hover {
  -webkit-transform: scale(1.3);
  -ms-transform: scale(1.3);
  transform: scale(1.3);
}

.c1 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #1785FB;
  -webkit-transition: left 0.1s ease,width 0.1s ease;
  transition: left 0.1s ease,width 0.1s ease;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  left: 0%;
  width: 0%;
}

.c0 {
  position: relative;
  padding: 8px 0 24px;
}

<div
  className="c0"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c1"
    disabled={false}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="c2"
      disabled={false}
      width="0%"
    />
    <div
      className="c3"
      disabled={false}
      draggable={undefined}
      onMouseDown={[Function]}
      onTouchMove={[Function]}
    >
      <span
        className="c4"
        disabled={false}
        tabIndex="0"
      />
      <span
        className="c5"
        disabled={false}
      >
        0
      </span>
    </div>
  </div>
</div>
`;

exports[`Range value 1`] = `
.c5 {
  color: #2C2C2C;
  font: 500 10px/12px 'Roboto';
  white-space: nowrap;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  display: inline-block;
  margin-top: 4px;
}

.c3 {
  left: 50%;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: absolute;
  -webkit-transition: left 0.1s ease;
  transition: left 0.1s ease;
  top: -6px;
  -webkit-transform: translate3d(-50%,0,0);
  -ms-transform: translate3d(-50%,0,0);
  transform: translate3d(-50%,0,0);
  padding: 0 0.5rem;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4 {
  -webkit-tap-highlight-color: transparent;
  -webkit-appearance: none;
  -moz-appearance: none;
  appearance: none;
  background: #fff;
  border-radius: 100%;
  border: 1px solid #1785FB;
  display: inline-block;
  height: 12px;
  outline: none;
  -webkit-transition: -webkit-transform 0.1s ease;
  -webkit-transition: transform 0.1s ease;
  transition: transform 0.1s ease;
  width: 12px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c4:hover {
  -webkit-transform: scale(1.3);
  -ms-transform: scale(1.3);
  transform: scale(1.3);
}

.c1 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.c2 {
  background: #E2E6E3;
  border-radius: 3px;
  cursor: pointer;
  height: 2px;
  position: relative;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  background: #1785FB;
  -webkit-transition: left 0.1s ease,width 0.1s ease;
  transition: left 0.1s ease,width 0.1s ease;
  -webkit-tap-highlight-color: transparent;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  left: 0%;
  width: 50%;
}

.c0 {
  position: relative;
  padding: 8px 0 24px;
}

<div
  className="c0"
  onMouseDown={[Function]}
  onTouchStart={[Function]}
>
  <div
    className="c1"
    disabled={false}
    onMouseDown={[Function]}
    onTouchStart={[Function]}
  >
    <div
      className="c2"
      disabled={false}
      width="50%"
    />
    <div
      className="c3"
      disabled={false}
      draggable={undefined}
      onMouseDown={[Function]}
      onTouchMove={[Function]}
    >
      <span
        className="c4"
        disabled={false}
        tabIndex="0"
      />
      <span
        className="c5"
        disabled={false}
      >
        20
      </span>
    </div>
  </div>
</div>
`;
